Data/ByteString/Internal/Type.hs

```diff
@@ -885,13 +885,36 @@ stimesPolymorphic nRaw !bs = case checkedIntegerToInt n of
   -- and the likelihood of potentially dangerous mistakes minimized.
 
 
+{-
+Note [Float error calls out of INLINABLE things]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If a function is marked INLINE or INLINABLE, then when ghc inlines or
+specializes it, it duplicates the function body exactly as written.
+
+This feature is useful for systems of rewrite rules, but sometimes
+comes at a code-size cost.  One situation where this cost generally
+comes with no compensating up-side is when the function in question
+calls `error` or something similar.
+
+Such an `error` call is not meaningfully improved by the extra context
+inlining or specialization provides, and if inlining or specialization
+happens in a different module from where the function was originally
+defined, CSE will not be able to de-duplicate the error call floated
+out of the inlined RHS and the error call floated out of the original
+RHS.  See also https://gitlab.haskell.org/ghc/ghc/-/issues/23823
+
+To mitigate this, we manually float the error calls out of INLINABLE
+functions when it is possible to do so.
+-}
+
 stimesNegativeErr :: ByteString
+-- See Note [Float error calls out of INLINABLE things]
 stimesNegativeErr
-  = error "stimes @ByteString: non-negative multiplier expected"
+  = errorWithoutStackTrace "stimes @ByteString: non-negative multiplier expected"
```

Comment:
Why no stack trace?

Reply:
The `Semigroup` class doesn't give its methods `HasCallStack`, so the trace will only show control flow internal to this module, which is useless to consumers.

Reply:
AFAIR you can put an explicit method signature with `HasCallStack`, in which case it will work out. I'm torn whether we want this.

Reply:
It will "work" in the sense that it compiles, but it does not help produce a useful stack trace. The fundamental problem is that the function stored in the `Semigroup ByteString` dictionary has type `forall b. Integral b => b -> ByteString -> ByteString` with no place for a use-site to provide `HasCallStack` information. See also [ghc ticket 22103](https://gitlab.haskell.org/ghc/ghc/-/issues/22103) and the relevant part of the [`InstanceSigs` documentation](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/instances.html#instance-signatures-type-signatures-in-instance-declarations):

> * The instance signature is purely local to the class instance declaration. It only affects the typechecking of the method in the instance; it does not affect anything outside the class instance. In this way, it is similar to an inline type signature:
> 
>   ```
>   instance Eq a => Eq (T a) where
>   
>       (==) = (x y -> True) :: forall b. b -> b -> Bool
>   ```
>   In particular, adding constraints such as _HasCallStack_ to the instance signature will not have an effect; they need to be added to the class instead.

Reply:
It seems documentation is wrong, it is possible to pass `HasCallStack` via `InstanceSigs`: https://github.com/haskell/core-libraries-committee/issues/115#issuecomment-1381544143

I'm fine with not threading call stack through `stimes` though.

Reply:
The documentation is not perfect; adding `HasCallStack` via `InstanceSigs` results in an extra `CallStack` entry that reads `a type signature in an instance, called at <location of the relevant instance signature>`. But it is morally correct; as [the comment directly preceding the one you linked to](https://github.com/haskell/core-libraries-committee/issues/115#issuecomment-1381222537) points out, "`HasCallStack` needs be on the typeclass decl to prevent the callstack chain from breaking, full stop."

Feel free to experiment with this yourself. It is not possible to thread `HasCallStack` information through `stimes` without changing the definition of `class Semigroup` in `base`. (Well, _maybe_ it's possible by using weird tricks to force `HasCallStack` to be part of the instance context, i.e. `instance DisguisedHasCallStack => Semigroup ByteString`. But GHC doesn't like that kind of thing generally.)

Reply:
Got it, feel free to merge.